{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\henry\\Anaconda3\\lib\\site-packages\\ipykernel_launcher.py:324: RuntimeWarning: divide by zero encountered in log2\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: tRNA-Gln-TTG-1 is not the expected length; skipping gapped regions\n",
      "Warning: tRNA-Gln-TTG-2 is not the expected length; skipping gapped regions\n",
      "Warning: tRNA-His-GTG-1 is not the expected length; skipping gapped regions\n",
      "Warning: tRNA-Leu-AAG-1 is not the expected length; skipping gapped regions\n",
      "Warning: tRNA-Ser-GGA-1 is not the expected length; skipping gapped regions\n",
      "Warning: tRNA-Ser-TGA-1 is not the expected length; skipping gapped regions\n",
      "Warning: tRNA-Undet-NNN-1 is not the expected length; skipping gapped regions\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Sort sequences and run CMs; generate output file\n",
    "\n",
    "1) filter out sequences with wrong base\n",
    "--> Run sprinzl positions; \n",
    "--> Output seqs with correct base to fasta\n",
    "2) Make predictions using cmsearch\n",
    "--> Parse cmsearch results from +/- files\n",
    "3) Sort cmsearch results into output file\n",
    "--> Write previously-specified output file OR matt's file format\n",
    "\n",
    "New edit: Parse through search fasta to find tRNA sequences that did not have any hits in the search results. \n",
    " - Make sure to account for them in some way\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def parseInput():\n",
    "    \"\"\"Parse command line input and output files\"\"\"\n",
    "    import argparse\n",
    "    \n",
    "    #Argparse setup\n",
    "    argParser = argparse.ArgumentParser(description = 'This program generates track hubs that display Modomics data')\n",
    "    argParser.add_argument('-s', '--tRNAscan_ss', required = True, help = 'list of tRNAscan-SE secondary structure file paths')\n",
    "    argParser.add_argument('-o', '--output_directory', required = True, help = 'Path to output results')\n",
    "    argParser.add_argument('-k', '--kingdom', required = True, choices = {'E', 'A', 'B'}, help = 'Domain of organism')\n",
    "    argParser.add_argument('--cm_database', required = False, default = './CMlibrary.txt', help = 'File containing information about tRNA mod CMs to search')\n",
    "    argParser.add_argument('--prob_database', required = False, default = './probLibrary.txt', help = 'File containing information about tRNA mod CMs to search')\n",
    "    argParser.add_argument('--skip_sprinzl_align', required = False, action = 'store_true', help = 'Skips sprinzl alignment step to save time if this has already been done.')\n",
    "    \n",
    "    kingdoms = {'E': 'Eukaryota', 'A': 'Archaea', 'B': 'Bacteria'}\n",
    "    \n",
    "    #'-s ../SlicerV2/data/secStruct/strePneu_TIGR4-tRNAs.ss.sort -o ./strepneumo_predictions -k B'.split()\n",
    "    clArgs = argParser.parse_args()\n",
    "    tRNAstruct = clArgs.tRNAscan_ss\n",
    "    orgKing = kingdoms[clArgs.kingdom]\n",
    "    outDir = clArgs.output_directory\n",
    "    cmFile = clArgs.cm_database\n",
    "    probFile = clArgs.prob_database\n",
    "    skipSprinzl = clArgs.skip_sprinzl_align\n",
    "    \n",
    "    return tRNAstruct, orgKing, outDir, cmFile, probFile, skipSprinzl\n",
    "\n",
    "def sortPositions(inputFile):\n",
    "    \"\"\"Read tRNA sprinzl position file\"\"\"\n",
    "    \n",
    "    sequence = str() #Store tRNA sequence\n",
    "    posDict = dict()  #Store tRNA sequence as a list\n",
    "    position = [] #Store sprinzl positions\n",
    "    \n",
    "    with open(inputFile, 'r') as inF:\n",
    "        lines = inF.readlines()\n",
    "        \n",
    "        for line in lines[2:]:\n",
    "            splitLine = line.strip().split('\\t')\n",
    "            \n",
    "            #Remove spaces for output fasta\n",
    "            if splitLine[1] != '-':\n",
    "                sequence += splitLine[1] #Full sequence\n",
    "            \n",
    "            posDict[splitLine[0]] = splitLine[1] #Dictionary of positions\n",
    "            position.append(splitLine[0]) #List of sprinzl positions\n",
    "        \n",
    "        inF.close()\n",
    "    \n",
    "    return sequence, position, posDict\n",
    "\n",
    "def writeFasta(outFileName, seqDict, blockSize = 60):\n",
    "    \"\"\"Write fasta ouput of unmodified tRNA sequences to use with tRNAscan-SE secondary structure\"\"\"\n",
    "    \n",
    "    with open(outFileName, 'w') as outF:\n",
    "        \n",
    "        for seqName in seqDict.keys():\n",
    "            \n",
    "            outF.write('>{0}\\n'.format(seqName))\n",
    "            \n",
    "            for block in range(0, len(seqDict[seqName]), blockSize):\n",
    "                try:\n",
    "                    outF.write('{0}\\n'.format(seqDict[seqName][block:block+60]))\n",
    "                except IndexError:\n",
    "                    outF.write('{0}\\n'.format(seqDict[seqName][block:]))\n",
    "        \n",
    "        outF.close()\n",
    "\n",
    "def cmSearchParser(inputFile):\n",
    "    \"\"\"Parse cmsearch --tblout file for e value and sequence name\"\"\"\n",
    "    \n",
    "    resultsDict = dict()\n",
    "    \n",
    "    with open(inputFile, 'r') as inF:\n",
    "        \n",
    "        lines = inF.readlines()\n",
    "        \n",
    "        #Iterate through lines in file\n",
    "        for line in lines:\n",
    "            \n",
    "            #Skip though metadata\n",
    "            if line[0] == '#':\n",
    "                pass\n",
    "            else:\n",
    "                splitLine = line.strip().split(' ')\n",
    "                \n",
    "                #Remove white space\n",
    "                values = []\n",
    "                for x in splitLine:\n",
    "                    if x != '':\n",
    "                        values.append(x)\n",
    "                    else:\n",
    "                        pass\n",
    "                \n",
    "                #Name useful info\n",
    "                name = values[0]\n",
    "                cm = values[1]\n",
    "                score = values[14]\n",
    "                E = values[15]\n",
    "                \n",
    "                resultsDict[name] = score\n",
    "        \n",
    "        inF.close()\n",
    "    \n",
    "    return resultsDict   \n",
    "\n",
    "def parseIsoacceptors(inFile):\n",
    "    \"\"\"Parse input file type for sorting tRNAs by modification state\"\"\"\n",
    "    \"\"\"Make sure to remove tRNAs from negative set if it appears in both + and - sets\"\"\"\n",
    "    \"\"\"Add option to specify specific slices in the tRNA sequence!!!\"\"\"\n",
    "    #Store data\n",
    "    \n",
    "    modomicsData = dict()\n",
    "    \n",
    "    with open(inFile, 'r') as inF:\n",
    "        \n",
    "        entries = inF.read().split('>')[1:] #Split on delimiter and skip empty first string\n",
    "        \n",
    "        for entry in entries:\n",
    "            \n",
    "            lines = entry.split('\\n')[:-1]\n",
    "            \n",
    "            #Parse info\n",
    "            info = lines[0].split('\\t')\n",
    "            \n",
    "            modomicsData[' '.join(info[0:2])] = dict() #Make a dictionary key for each modified position in file\n",
    "            \n",
    "            #Store unmodified base and mod position\n",
    "            modomicsData[' '.join(info[0:2])]['refBase'] = info[2]\n",
    "            modomicsData[' '.join(info[0:2])]['refPos'] = info[1]\n",
    "        \n",
    "            for line in lines[1:]:\n",
    "                splitLine = line.split('\\t')\n",
    "                \n",
    "                #Additional sorting criteria to make parsing easier later on\n",
    "                species = splitLine[0].split('-')[0]\n",
    "                isoacceptor = '-'.join(splitLine[0].split('-')[1:4])\n",
    "                \n",
    "                #Sort entries in sets\n",
    "                try:\n",
    "                    try:\n",
    "                        modomicsData[' '.join(info[0:2])][species][splitLine[1]].add(isoacceptor)\n",
    "                    except KeyError:\n",
    "                        modomicsData[' '.join(info[0:2])][species] = {'+': set(), '-': set(), 'u': set()}\n",
    "                        modomicsData[' '.join(info[0:2])][species][splitLine[1]].add(isoacceptor)\n",
    "                except KeyError:\n",
    "                    modomicsData[' '.join(info[0:2])][species] = {'+': set(), '-': set(), 'u': set()}\n",
    "                    modomicsData[' '.join(info[0:2])][species][splitLine[1]].add(isoacceptor)\n",
    "                    \n",
    "        inF.close()\n",
    "    \n",
    "    return modomicsData   \n",
    "\n",
    "def parseSummary(inputFile):\n",
    "    \"\"\"Parse input summary file for CMs\"\"\"\n",
    "    \n",
    "    infoDict = dict()\n",
    "    modifiedPositions = dict()\n",
    "    \n",
    "    with open(inputFile, 'r') as inF:\n",
    "        lines = inF.readlines()[1:]\n",
    "        \n",
    "        for line, num in zip(lines, range(0, len(lines))):\n",
    "            \n",
    "            splitLine = line.strip().split('\\t')\n",
    "            \n",
    "            infoDict['_'.join([splitLine[0], splitLine[1], splitLine[2]])] = {\n",
    "                             'mod': splitLine[0], \n",
    "                             'pos': splitLine[1], \n",
    "                             'refBase': splitLine[2], \n",
    "                             'kingdom': splitLine[3], \n",
    "                             'posCM': splitLine[4], \n",
    "                             'negCM': splitLine[5]}\n",
    "            try:\n",
    "                modifiedPositions['_'.join([splitLine[1], splitLine[2]])] += '_'.join([splitLine[0], splitLine[1], splitLine[2]])\n",
    "            except KeyError:\n",
    "                modifiedPositions['_'.join([splitLine[1], splitLine[2]])] = ['_'.join([splitLine[0], splitLine[1], splitLine[2]])]\n",
    "            \n",
    "        inF.close()\n",
    "    \n",
    "    return infoDict, modifiedPositions\n",
    "\n",
    "def nameIsodecoders(inputFile):\n",
    "    \"\"\"\n",
    "    Read in tRNA isodecoder names and output dictionary of named isodecoders with their sequences and corresponding gene names\n",
    "    Note: WONT HANDLE INTRONS!!!!!\n",
    "    \"\"\"\n",
    "    \n",
    "    acceptorScores = {}\n",
    "    acceptorNames = {}\n",
    "    \n",
    "    with open(inputFile, 'r') as inF:\n",
    "        \n",
    "        entries = inF.read().split('\\n\\n')\n",
    "        \n",
    "        for entry in entries[:-1]:\n",
    "            tRNAinfo = entry.split('\\n')\n",
    "            \n",
    "            #tRNAscan ID name\n",
    "            tRNAscanID = tRNAinfo[0].split(' ')[0]\n",
    "            \n",
    "            #Name isoacceptor\n",
    "            AA = tRNAinfo[1].split('\\t')[0].split(' ')[-1] #Amino acid\n",
    "            AC = tRNAinfo[1].split(' ')[2] #Anticodon\n",
    "            isoacceptorName = '-'.join([AA, AC])\n",
    "            \n",
    "            #Score\n",
    "            score = float(tRNAinfo[1].split(' ')[-1])\n",
    "            \n",
    "            #sequence\n",
    "            sequence = tRNAinfo[-2].split(' ')[-1]\n",
    "            \n",
    "            try:\n",
    "                try:\n",
    "                    acceptorScores[isoacceptorName][sequence] = score\n",
    "                    acceptorNames[isoacceptorName][sequence].append(tRNAscanID)\n",
    "                except KeyError:\n",
    "                    acceptorScores[isoacceptorName][sequence] = score\n",
    "                    acceptorNames[isoacceptorName][sequence] = [tRNAscanID]\n",
    "            except KeyError:\n",
    "                acceptorScores[isoacceptorName] = {sequence: score}\n",
    "                acceptorNames[isoacceptorName] = {sequence: [tRNAscanID]}\n",
    "    \n",
    "    isoacceptors = {}\n",
    "    isoFastaDict = {}\n",
    "    scanNameToIsoName = {}\n",
    "    \n",
    "    #Sort isodecoders by score and name them\n",
    "    for acc in acceptorScores.keys():\n",
    "        for tRNA, pos in zip(acceptorScores[acc].keys(), range(1, len(acceptorScores[acc].keys())+1, 1)):\n",
    "            \n",
    "            for score in sorted(acceptorScores[acc].values()):\n",
    "                \n",
    "                if acceptorScores[acc][tRNA] == score:\n",
    "                    tRNAnames = [str(x) for x in acceptorNames[acc][tRNA]]\n",
    "                    faHeader = '{0}\\tScore: {1}\\tGene names: {2}'.format('-'.join([str(acc), str(pos)]), str(score), ', '.join(tRNAnames))\n",
    "                    isoName = '-'.join([acc, str(pos)])\n",
    "                    isoacceptors[isoName] = {'seq': tRNA, \n",
    "                                             'gene names': acceptorNames[acc][tRNA], \n",
    "                                             'score': score}\n",
    "                    isoFastaDict[tRNA.upper()] = isoName\n",
    "                    \n",
    "                    for scanName in acceptorNames[acc][tRNA]:\n",
    "                        scanNameToIsoName[scanName] = isoName\n",
    "                    \n",
    "                else:\n",
    "                    pass\n",
    "    \n",
    "    return isoacceptors, isoFastaDict, scanNameToIsoName\n",
    "    \n",
    "\n",
    "\n",
    "def parseProbs(inputFile):\n",
    "    \"\"\"Parse conditional probability input file\"\"\"\n",
    "    \n",
    "    #Store conditional probabilities\n",
    "    probsDict = {}\n",
    "    \n",
    "    with open(inputFile, 'r') as inF:\n",
    "        lines = inF.readlines()\n",
    "        \n",
    "        for line in lines[1:]:\n",
    "            splitLine = line.strip().split('\\t')\n",
    "            \n",
    "            modLabel = '_'.join(splitLine[0:3])\n",
    "            residue = splitLine[3]\n",
    "            pos = splitLine[4]\n",
    "            posProb = float(splitLine[5])\n",
    "            negProb = float(splitLine[6])\n",
    "            \n",
    "            #Compile dictionary\n",
    "            try:\n",
    "                try:\n",
    "                    probsDict[modLabel][pos][residue] = [posProb, negProb]\n",
    "                except KeyError:\n",
    "                    probsDict[modLabel][pos] = {residue: [posProb, negProb]}\n",
    "                        \n",
    "            except KeyError:\n",
    "                probsDict[modLabel] = {pos: {residue: [posProb, negProb]}}\n",
    "                \n",
    "        inF.close()\n",
    "    \n",
    "    return probsDict\n",
    "    \n",
    "def assignProb(tRNAseqs, modInfo, probDict):\n",
    "    \"\"\"Calculate conditional probabilities for anticodon modifications\"\"\"\n",
    "    import numpy as np\n",
    "    mod, refPos, refBase = modInfo.split('_')\n",
    "    \n",
    "    probs = {'modScore':{}}\n",
    "    \n",
    "    #Iterate through sequences\n",
    "    for iso in tRNAseqs.keys():\n",
    "        \n",
    "        seq = tRNAseqs[iso]\n",
    "        \n",
    "        #Iterate through interesting positions\n",
    "        #for pos in probDict.keys(): ##################################################################\n",
    "        #Figure out how to combine if multiple positions specified!!!!\n",
    "        pos = list(probDict.keys())[0]\n",
    "        \n",
    "        #Handle U vs T in different datasets\n",
    "        seqBase = seq[pos]\n",
    "        if seq[pos] == 'T':\n",
    "            seqBase = 'U'\n",
    "        \n",
    "        try:\n",
    "            #Get conditional probabilities for observed base\n",
    "            probs['modScore'][iso] = np.log2(probDict[pos][seqBase][0]/probDict[pos][seqBase][1])\n",
    "            \n",
    "        except KeyError:\n",
    "            if seqBase == '-':\n",
    "                pass\n",
    "            else:\n",
    "                print(pos, seqBase)\n",
    "          \n",
    "    return probs\n",
    "        \n",
    "def main(inCl = True, ssFile = None, domain = None, outputFile = None, cmsToSearch = './CMlibrary.txt', condProbs = './probLibrary.txt', skipSpr = False):\n",
    "    \"\"\"Execute commands\"\"\"\n",
    "    import os\n",
    "    \n",
    "    #Recieve command line args\n",
    "    #tRNAstruct, orgKing, outfile, sortFile , runFile\n",
    "    if inCl:\n",
    "        ssFile, domain, outputFile, cmsToSearch, condProbs, skipSpr = parseInput()\n",
    "    \n",
    "    tempDir = '{0}/tmp'.format('/'.join(outputFile.split('/')))\n",
    "    \n",
    "    #Directory for output files \n",
    "    try:\n",
    "        os.system('mkdir {0}'.format('/'.join(outputFile.split('/'))))\n",
    "    except FileExistsError:\n",
    "        pass\n",
    "    \n",
    "    #Make temporary directory for sprinzl position output\n",
    "    try:\n",
    "        os.system('mkdir {0}'.format(tempDir))\n",
    "    except FileExistsError:\n",
    "        pass\n",
    "    \n",
    "    #Parse CM summary file\n",
    "    allCMs, modInfo = parseSummary(cmsToSearch)\n",
    "    \n",
    "    #Parse conditional probability file\n",
    "    probsDict = parseProbs(condProbs)\n",
    "    #Add to modInfo\n",
    "    for modTag in probsDict.keys():\n",
    "        try:\n",
    "            modInfo['_'.join(modTag.split('_')[1:])].append(modTag)\n",
    "        except KeyError:\n",
    "            modInfo['_'.join(modTag.split('_')[1:])] = [modTag]\n",
    "    \n",
    "    #Store tRNA sequences\n",
    "    tRNAseq = dict() #Store tRNA sequences {tRNAscan-SE ID: sequence}\n",
    "    tRNApos = dict() #Store tRNA positions {tRNAscan-SE ID:{sprinzl position: base}}\n",
    "    isoDict = dict() #Store isodecoder sequences and respective tRNAscan-SE IDs {sequence: [tRNAscan-SE ID]}\n",
    "    \n",
    "    #Store prediction statistics\n",
    "    predStats = dict()\n",
    "    \n",
    "    #Store tRNA isodecoder names\n",
    "    faDict = {}\n",
    "    isoPosDict = {}\n",
    "    \n",
    "    isoInfo, fastaSeqs, nameMap = nameIsodecoders(ssFile)\n",
    "        \n",
    "    outD = '{0}/sprinzl_alignments'.format(tempDir)\n",
    "        \n",
    "    if skipSpr == False:\n",
    "        try:\n",
    "            os.system('mkdir {0}'.format(outD))\n",
    "            os.system('./tRNA_sprinzl_pos -c ./map-sprinzl-pos.conf -d {0} -s {1} -o {2}'.format(domain, ssFile, outD))\n",
    "        except FileExistsError:\n",
    "            os.system('./tRNA_sprinzl_pos -c ./map-sprinzl-pos.conf -d {0} -s {1} -o {2}'.format(domain, ssFile, outD))\n",
    "\n",
    "    #Read through files from sprinzl position namer.\n",
    "    sprinzlFiles = os.listdir(outD)\n",
    "        \n",
    "    #Store CM results for each mod\n",
    "    predResults = {}\n",
    "        \n",
    "    #Iterate through specified CMs:\n",
    "    for cm in allCMs.keys():\n",
    "        \n",
    "        refBase = allCMs[cm]['refBase']\n",
    "        refPos = allCMs[cm]['pos']\n",
    "        refMod = allCMs[cm]['mod']\n",
    "        \n",
    "        predResults['_'.join([refMod, refPos, refBase])] = {}\n",
    "        \n",
    "        #Iterate through sprinzl alignments\n",
    "        for file in sprinzlFiles:\n",
    "\n",
    "            #Define relevant file names\n",
    "            fName = '{0}/{1}'.format(outD, file)\n",
    "            tRNAscanID = '.'.join(file.split('.')[:-1])\n",
    "\n",
    "            #Skip tmp files\n",
    "            if fName.split('.')[-1] == 'pos':\n",
    "\n",
    "                #Read file\n",
    "                sequence, position, posDict = sortPositions(fName)\n",
    "\n",
    "                if posDict[refPos] == refBase:\n",
    "                    tRNAseq[tRNAscanID] = sequence\n",
    "                    tRNApos[tRNAscanID] = posDict\n",
    "\n",
    "                    #Sort by isodecoders:\n",
    "                    try:\n",
    "                        isoDict[sequence].append(tRNAscanID)\n",
    "                    except KeyError:\n",
    "                        isoDict[sequence] = [tRNAscanID]\n",
    "                else:\n",
    "                    pass\n",
    "            else:\n",
    "                pass\n",
    "        \n",
    "        #Re-pair sequences with isoacceptor names\n",
    "        for seq in isoDict.keys():\n",
    "            for ID in isoDict[seq]:\n",
    "                faDict[nameMap[ID.split('-')[0]]] = seq\n",
    "                isoPosDict[nameMap[ID.split('-')[0]]] = tRNApos[ID]\n",
    "        \n",
    "        #Write modifiable sequences into fasta file for cmSearch\n",
    "        faName = '{0}/{1}_{2}.fasta'.format(tempDir, refPos, refBase)\n",
    "        writeFasta(faName, faDict)\n",
    "            \n",
    "        #Search sequences against fasta file\n",
    "        os.system('cmsearch -g --tblout {0}-{1}-hits.txt {2} {0}'.format(faName, allCMs[cm]['posCM'].split('/')[-1], allCMs[cm]['posCM']))\n",
    "        os.system('cmsearch -g --tblout {0}-{1}-hits.txt {2} {0}'.format(faName, allCMs[cm]['negCM'].split('/')[-1], allCMs[cm]['negCM']))\n",
    "            \n",
    "        #Handle when there is no results file (in the case that one CM has no sequences in it)\n",
    "        try:\n",
    "            predResults['_'.join([refMod, refPos, refBase])]['posResults'] = cmSearchParser('{0}-{1}-hits.txt'.format(faName, allCMs[cm]['posCM'].split('/')[-1]))\n",
    "        except FileNotFoundError:\n",
    "            print('Warning: positive search results file not found')\n",
    "            posResults = dict()\n",
    "                \n",
    "        try:\n",
    "            predResults['_'.join([refMod, refPos, refBase])]['negResults'] = cmSearchParser('{0}-{1}-hits.txt'.format(faName, allCMs[cm]['negCM'].split('/')[-1]))\n",
    "        except FileNotFoundError:\n",
    "            print('Warning: negative search results file not found')\n",
    "            negResults = dict()\n",
    "        \n",
    "        #Add anticodon-loop conditional probabilities\n",
    "        \n",
    "        for modTag in probsDict.keys():            \n",
    "            predResults[modTag] = assignProb(isoPosDict, modTag, probsDict[modTag])\n",
    "        \n",
    "    \n",
    "    #Output file labelling\n",
    "            \n",
    "    callFile = '{0}/modCalls.txt'.format(outputFile)\n",
    "    scoreFile = '{0}/modScores.txt'.format(outputFile)\n",
    "    \n",
    "    #write output file\n",
    "    writeOutput(callFile, scoreFile, predResults, tRNApos, allCMs, modInfo, nameMap)\n",
    "    \n",
    "    return callFile, scoreFile\n",
    "\n",
    "def writeOutput(callsFile, scoreFile, preds, isos, searchedMods, modPositions, nameMap):\n",
    "    \"\"\"Sort isodecoders based on MODOMICS input file\"\"\"\n",
    "    from tRNAinfo import sprinzl2coords as s2c\n",
    "    \n",
    "    with open(callsFile, 'w') as cF, open(scoreFile, 'w') as sF:\n",
    "        \n",
    "        sprinzlPos = list(s2c.keys())[1:]\n",
    "        \n",
    "        cF.write('isoacceptor\\t{0}\\n'.format('\\t'.join(sprinzlPos)))\n",
    "        sF.write('isoacceptor\\t{0}\\n'.format('\\t'.join(sprinzlPos)))\n",
    "        \n",
    "        #reorder into isodecoder names\n",
    "        isoAlign = {}\n",
    "        for tRNA in isos.keys():\n",
    "            isoAlign[nameMap[tRNA.split('-')[0]]] = isos[tRNA]\n",
    "        \n",
    "        #Write results to file\n",
    "        for iso in sorted(isoAlign.keys()):\n",
    "            cF.write('tRNA-{0}'.format(iso))\n",
    "            sF.write('tRNA-{0}'.format(iso))\n",
    "            \n",
    "            #Warn about abnormal length tRNAs:\n",
    "            if len(sprinzlPos) != len(isoAlign[iso].keys()):\n",
    "                print('Warning: tRNA-{0} is not the expected length; skipping gapped regions'.format(iso))\n",
    "            \n",
    "            #Iterate through sprinzl positions\n",
    "            for pos in sprinzlPos:\n",
    "                \n",
    "                base = isoAlign[iso][pos]\n",
    "                \n",
    "                #calc mod odds score, call mod/unmod based on odds score\n",
    "                try:\n",
    "                    scoreStr = []\n",
    "                    modScores = [] #Store multiple scores if present\n",
    "                    modList = []\n",
    "                    for mod in modPositions['_'.join([pos, base])]:\n",
    "                        modScore = float()\n",
    "                        #calculate mod score\n",
    "                        if 'modScore' in preds[mod].keys():\n",
    "                            modScore = preds[mod]['modScore'][iso]\n",
    "                            #scoreStr.append(str(modScore))\n",
    "                        else:\n",
    "                            modScore = float(preds[mod]['posResults'][iso]) - float(preds[mod]['negResults'][iso])\n",
    "                        \n",
    "                        #get short name from sorting string\n",
    "                        shortName = mod.split('_')[0]\n",
    "                        \n",
    "                        #Store multiple mod scores\n",
    "                        modScores.append(modScore)\n",
    "                        modList.append(shortName)\n",
    "                   \n",
    "                    ###################################################\n",
    "                    #Handle when multiple mods can occur at the same position -> Select the mod that has the highest score\n",
    "                    for mod, score in zip(modList, modScores):\n",
    "                        \n",
    "                        if score == max(modScores): #If two have the same count, it writes both...\n",
    "                            \n",
    "                            if score > 0:\n",
    "                                \n",
    "                                cF.write('\\t{0}'.format(mod))\n",
    "                                sF.write('\\t{0}'.format(round(score, 2)))\n",
    "                                break\n",
    "\n",
    "                            else:\n",
    "                                cF.write('\\t{0}'.format(base))\n",
    "                                sF.write('\\t{0}'.format(round(score, 2)))\n",
    "                                break\n",
    "                        else:\n",
    "                            pass \n",
    "                    \n",
    "                \n",
    "                #Write reference base if no modification available\n",
    "                except KeyError:\n",
    "                    cF.write('\\t{0}'.format(base)) #Write base at unmodified position\n",
    "                    sF.write('\\t-') #Denote no mod prediction at this spot\n",
    "                    \n",
    "            cF.write('\\n')\n",
    "            sF.write('\\n')\n",
    "                \n",
    "        cF.close()\n",
    "        sF.close()\n",
    "    \n",
    "main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
